rest/api.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.  // You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package rest
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus-community/promql-langserver/internal/vendored/go-tools/lsp/protocol"
+	"github.com/prometheus-community/promql-langserver/langserver"
+	promClient "github.com/prometheus-community/promql-langserver/prometheus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/route"
+)
+
+func returnJSON(w http.ResponseWriter, content interface{}) {
+	encoder := json.NewEncoder(w)
+
+	err := encoder.Encode(content)
+	if err != nil {
+		http.Error(w, errors.Wrapf(err, "failed to write response").Error(), 500)
+	}
+}
+
+type lspData struct {
+	// Expr is the promQL expression and is required for all endpoint available.
+	Expr string `json:"expr"`
+	// Limit is the number max of result returned to the client. It will be used for the autocompletion and the diagnostics.
+	Limit *uint64 `json:"limit,omitempty"`
+	// PositionLine is the number of the line for which the metadata is queried.
+	PositionLine *float64 `json:"positionLine,omitempty"`
+	// PositionChar for which the metadata is queried. Characters are counted as UTF16 Codepoints.
+	PositionChar *float64 `json:"positionChar,omitempty"`
+}
+
+func (d *lspData) UnmarshalJSON(data []byte) error {
+	var tmp lspData
+	type plain lspData
+	if err := json.Unmarshal(data, (*plain)(&tmp)); err != nil {
+		return err
+	}
+	if len(tmp.Expr) == 0 {
+		return fmt.Errorf("promQL expression is not specified")
+	}
+	*d = tmp
+	return nil
+}
+
+func (d *lspData) returnPosition() (protocol.Position, error) {
+	if d.PositionLine == nil {
+		return protocol.Position{}, errors.New("positionLine is not specified")
+	}
+	if d.PositionChar == nil {
+		return protocol.Position{}, errors.New("positionChar is not specified")
+	}
+	return protocol.Position{
+		Line:      *d.PositionLine,
+		Character: *d.PositionChar,
+	}, nil
+}
+
+type API struct {
```

Comment:
Is this struct intended to satisfy a specific interface?

If yes, a value with the interface type should be returned by `NewLangserver`, to ensure it is actually implemented..

Reply:
No this struct doesn't implement any kind of interface.
It's just a struct to handle the different endpoints and how they are manages/register

Reply:
Then there should still be a doc comment explaining what this object does.

Reply:
> If yes, a value with the interface type should be returned by NewLangserver, to ensure it is actually implemented..

Btw., even if not applicable here anyway, generally it is encouraged to return concrete implementation types: https://github.com/golang/go/wiki/CodeReviewComments#interfaces, "The implementing package should return concrete (usually pointer or struct) types: that way, new methods can be added to implementations without requiring extensive refactoring."

If you want to just test that a type implements an interface, in practice people either just leave that up to the caller / usage site (if that's within the same CI testing realm), or by just saying `var _ MyInterface = &MyImplementation{}` somewhere next to their implementation.